common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -539,134 +522,117 @@ public static String bytesToIpAddress(byte[] bytes, int offset, int length) {
+                colons++;
+                wordStart = curIdx + 1;
+                wordLen = 0;
+                break;
+            case '%':
+                // Strip the interface name/index after the percent sign
+                break loop;
+            default:
+                if (wordLen > 3 || !isValidHexChar(c)) {
+                    return false;
+                }
+                wordLen++;
+                prevColon = false;
             }
         }
 
         // Check if we have an IPv4 ending
-        if (numberOfPeriods > 0) {
-            // There is a test case with 7 colons and valid ipv4 this should resolve it
-            if (numberOfPeriods != 3 || !(isValidIp4Word(word.toString()) && numberOfColons < 7)) {
-                return false;
-            }
-        } else {
-            // If we're at then end and we haven't had 7 colons then there is a
-            // problem unless we encountered a doubleColon
-            if (numberOfColons != 7 && !doubleColon) {
-                return false;
-            }
-
-            // If we have an empty word at the end, it means we ended in either
-            // a : or a .
-            // If we did not end in :: then this is invalid
-            if (word.length() == 0 && ipAddress.charAt(length - 1 - startOffset) == ':' &&
-                ipAddress.charAt(length - 2 - startOffset) != ':') {
-                return false;
-            }
+        if (dots > 0) {
+            return dots == 3 &&
+                   // normal case (6 single colons) or a special case with double colon
+                   // such as ::1:2:3:4:5:d.d.d.d or ::ffff:5.6.7.8 allows to 7 colons with an IPv4 ending
+                   ((colons == 6 && !doubleColon) || (colons <= 7 && colonsAtStart)) &&
+                   // validate last word
+                   isValidIpV4Word(ip, wordStart, wordStart + wordLen);
         }
 
-        return true;
+        // If we're at then end and we haven't had 7 colons then there is a
+        // problem unless we encountered a doubleColon
+        return (colons == 7 || doubleColon) &&
+               // If we have an empty word at the end, it means we ended in either a : or a .
+               // If we did not end in :: then this is invalid
+               (wordLen > 0 || (prevColon && doubleColon));
     }
 
-    private static boolean isValidIp4Word(String word) {
-        char c;
-        if (word.length() < 1 || word.length() > 3) {
+    private static boolean isValidIpV4Word(CharSequence word, int start, int endExclusive) {
+        int len = endExclusive - start;
+        char c0, c1, c2;
+        if (len < 1 || len > 3 || (c0 = word.charAt(start)) < '0') {
             return false;
         }
-        for (int i = 0; i < word.length(); i ++) {
-            c = word.charAt(i);
-            if (!(c >= '0' && c <= '9')) {
-                return false;
-            }
+        if (len == 3) {
+            return (c1 = word.charAt(start + 1)) >= '0' &&
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of conditional operators (9) used in the expression (maximum allowed 5). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1067)